### ISSUE TITLE
New functions and operations for working with arrays

### ISSUE THREAD

**Issue by izveigor** [apache/arrow-datafusion#6119]
### Is your feature request related to a problem or challenge?

Arrow-DataFusion can support creating an array of a certain data type (`make_array`), but there are no functions to work with them.
In my opinion, the documentation of the database PostgreSQL can serve as a reference for working on arrays.

The following features are considered for implementation:
Operators:
- [ ] Any

Array operators:
- [ ] concatenation
- [ ] overlap
- [ ] concatenate

Array functions:
- [ ] array_append
- [ ] array_cat
- [ ] array_ndims
- [ ] array_dims
- [ ] array_fill
- [ ] array_length
- [ ] array_lower
- [ ] array_position
- [ ] array_positions
- [ ] array_prepend
- [ ] array_remove
- [ ] array_replace
- [ ] array_to_string
- [ ] array_upper
- [ ] cardinality
- [ ] string_to_array
- [ ] unnest

### Describe the solution you'd like

Series of PR, active discussion of features.

### Describe alternatives you've considered

The significance of some features can be disproven as a result of conversations.

### Additional context

Sources:
The array documentation: [PostgreSQL](https://postgrespro.ru/docs/postgrespro/15/functions-array?lang=en)
#6073
#6075

**Comment by mustafaakin** [comment]
We have a lot of nested data at [Resmo](www.resmo.com) and currently using [PartiQL](www.partiql.org) but we were exploring to switch to data-fusion. We make use a lot of *unnest*.

**Comment by alamb** [comment]
Thank you @izveigor  -- I have added this ticket to the tracking epic for working with lists in DataFusion: https://github.com/apache/arrow-datafusion/issues/2326

**Comment by BubbaJoe** [comment]
@alamb Isn't there still some functions/operators that need to be finished?

**Comment by alamb** [comment]
I agree -- thanks @BubbaJoe . @izveigor  FYI I reopened this one. Perhaps it would be good to file a follow on ticket with what is not yet implemented and then we can close this ticket
